lib/backend/k8s/resources/profile.go

```diff
@@ -188,42 +191,81 @@ func (c *profileClient) List(ctx context.Context, list model.ListInterface, revi
 		return nil, err
 	}
 
-	// Otherwise, enumerate all.
-	namespaces, err := c.clientSet.CoreV1().Namespaces().List(ctx, metav1.ListOptions{ResourceVersion: nsRev})
-	if err != nil {
-		return nil, K8sErrorToCalico(err, nl)
+	// Enumerate all namespaces, paginated.
+	listFunc := func(ctx context.Context, opts metav1.ListOptions) (runtime.Object, error) {
+		return c.clientSet.CoreV1().Namespaces().List(ctx, opts)
 	}
-
-	// For each Namespace, return a profile.
-	for _, ns := range namespaces.Items {
-		kvp, err := c.getNsKv(&ns)
+	forEach := func(obj runtime.Object) error {
+		ns := obj.(*v1.Namespace)
+		kvp, err := c.getNsKv(ns)
 		if err != nil {
+			// Return nil so we don't let one error interrupt processing of other objs.
 			log.Errorf("Unable to convert k8s Namespace to Calico Profile: Namespace=%s: %v", ns.Name, err)
-			continue
+			return nil
 		}
 		kvps = append(kvps, kvp)
+		return nil
+	}
+	lp := pager.New(listFunc)
+	opts := metav1.ListOptions{ResourceVersion: nsRev}
+	if nsRev != "" {
+		opts.ResourceVersionMatch = metav1.ResourceVersionMatchNotOlderThan
+	}
+	result, _, err := lp.List(ctx, opts)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, list)
+	}
+	err = meta.EachListItem(result, forEach)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, list)
 	}
 
-	// Enumerate all SA
-	var serviceaccounts *kapiv1.ServiceAccountList
-	// TBD: narrow down to only to the required namespace
-	serviceaccounts, err = c.clientSet.CoreV1().ServiceAccounts(kapiv1.NamespaceAll).List(ctx, metav1.ListOptions{ResourceVersion: saRev})
+	// Extract the list revision information.
+	m, err := meta.ListAccessor(result)
 	if err != nil {
-		return nil, K8sErrorToCalico(err, nl)
+		return nil, err
 	}
+	latestNSRev := m.GetResourceVersion()
 
-	for _, sa := range serviceaccounts.Items {
-		kvp, err := c.getSaKv(&sa)
+	// Enumerate all service accounts, paginated.
+	listFunc = func(ctx context.Context, opts metav1.ListOptions) (runtime.Object, error) {
+		return c.clientSet.CoreV1().ServiceAccounts(v1.NamespaceAll).List(ctx, opts)
+	}
+	forEach = func(obj runtime.Object) error {
+		sa := obj.(*v1.ServiceAccount)
+		kvp, err := c.getSaKv(sa)
 		if err != nil {
 			log.WithError(err).Errorf("Unable to convert k8s service account to Calico Profile: %s", sa.Name)
-			continue
+			return nil
 		}
 		log.Debug("Converted k8s sa to Calico profile ", sa.Name)
 		kvps = append(kvps, kvp)
+		return nil
+	}
+	lp = pager.New(listFunc)
+	opts = metav1.ListOptions{ResourceVersion: saRev}
+	if saRev != "" {
+		opts.ResourceVersionMatch = metav1.ResourceVersionMatchNotOlderThan
 	}
+	result, _, err = lp.List(ctx, opts)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, list)
+	}
+	err = meta.EachListItem(result, forEach)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, list)
+	}
+
+	// Extract the list revision information.
+	m, err = meta.ListAccessor(result)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Yes, it definitely feels now like we need a common utility for this passage of code.

Reply:
I can do a common utility - but I think it conflicts with the idea of inlining the `forEach` and `listFunc` functions - those will vary for each client and will need to be passed in to the common utility.